### PR TITLE
Gradient checkpointing

### PR DESCRIPTION
**Pull request by neverix**
This patch enables gradient checkpointing for ruDALLE.

It's possible to use up to 3x higher batch sizes in memory-limited environments during training.

Setting the `gradient_checkpointing` during `model.forward` makes a checkpoint every `gradient_checkpointing` layers.
6 is a good starting value.

### CODE REVIEW

**Comment by neverix** [comment]
:(

**Comment by shonenkov** [comment]
@neverix  we have a gradient checkpointing in the main training pipeline, but your version doesn't support deepspeed - so I can't merge your code right now - I beware of lost compatibility of our internal and open-source models :( 

I suggest creating a new branch with your version gradient checkpointing, what about it?

**Comment by neverix** [comment]
What changes are needed for compatibility with DeepSpeed? Having this included in the default install is useful for notebooks, if it's not possible we should look for different solutions
